src/librustc_lint/builtin.rs

```diff
@@ -591,6 +614,28 @@ impl Deprecated {
             cx.span_lint(lint, span, &msg);
         }
     }
+
+    fn push_item(&mut self, cx: &LateContext, item_id: ast::NodeId) {
+        self.def_stack.push(cx.tcx.map.local_def_id(item_id).index);
+    }
+
+    fn item_post(&mut self, cx: &LateContext, item_id: ast::NodeId) {
+        let top = self.def_stack.pop();
+
+        if let Some(id) = top {
+            if DefId::local(id) == cx.tcx.map.local_def_id(item_id) { return; }
+        }
+
+        bug!("unexpected lint traversal order: expected  `_post` for {:?} but called for {:?}",
+             top, item_id);
+    }
```

Comment:
Remember that all the lints have the cost of a single visitor; they are threaded together. This cannot be done with arbitrary visitors. The lint hooks exist basically for this use case.

That said, can't we pull instead of push here? We can use parent nodes and walk up the node tree. The case of a lint being triggered is relatively rare, so paying the cost of a tree walk there isn't too bad.